File: crates/kino-python/examples/basic_analysis.py
#!/usr/bin/env python3
"""
Basic Frequency Analysis Example

This example demonstrates how to use the psm_frequency Python bindings
to analyze audio files and extract frequency information.

Usage:
    python basic_analysis.py audio.wav
    python basic_analysis.py video.mp4
"""

import sys
import numpy as np
from kino_frequency import FrequencyAnalyzer, Fingerprinter, ContentTagger


def analyze_audio_file(file_path: str) -> None:
    """Perform basic frequency analysis on an audio/video file."""

    # Initialize the analyzer with 44.1kHz sample rate
    analyzer = FrequencyAnalyzer(sample_rate=44100)

    print(f"Analyzing: {file_path}")
    print("-" * 50)

    # Analyze the file
    result = analyzer.analyze(file_path)

    # Display dominant frequencies
    print("\nDominant Frequencies:")
    for i, freq in enumerate(result.dominant_frequencies[:5], 1):
        print(f"  {i}. {freq.frequency:.1f} Hz (magnitude: {freq.magnitude:.2f})")

    # Display band energies
    print("\nBand Energies:")
    bands = result.band_energies
    print(f"  Sub-bass (20-60 Hz):    {bands.sub_bass:.4f}")
    print(f"  Bass (60-250 Hz):       {bands.bass:.4f}")
    print(f"  Low-mid (250-500 Hz):   {bands.low_mid:.4f}")
    print(f"  Mid (500-2000 Hz):      {bands.mid:.4f}")
    print(f"  High-mid (2-4 kHz):     {bands.high_mid:.4f}")
    print(f"  Presence (4-6 kHz):     {bands.presence:.4f}")
    print(f"  Brilliance (6-20 kHz):  {bands.brilliance:.4f}")

    # Display spectral features
    print("\nSpectral Features:")
    print(f"  Centroid:  {result.spectral_centroid:.1f} Hz")
    print(f"  Rolloff:   {result.spectral_rolloff:.1f} Hz")
    print(f"  Flatness:  {result.spectral_flatness:.4f}")
    print(f"  ZCR:       {result.zero_crossing_rate:.4f}")


def generate_fingerprint(file_path: str) -> None:
    """Generate an audio fingerprint for content verification."""

    fingerprinter = Fingerprinter()

    print(f"\nGenerating fingerprint for: {file_path}")
    print("-" * 50)

    fingerprint = fingerprinter.fingerprint(file_path)

    print(f"  Hash:        {fingerprint.hash[:32]}...")
    print(f"  Duration:    {fingerprint.duration_secs:.2f} seconds")
    print(f"  Peak count:  {fingerprint.peak_count}")
    print(f"  Sample rate: {fingerprint.sample_rate} Hz")


def auto_tag_content(file_path: str) -> None:
    """Automatically generate content tags based on audio characteristics."""

    tagger = ContentTagger()

    print(f"\nAuto-tagging: {file_path}")
    print("-" * 50)

    tags = tagger.predict(file_path)

    print("  Predicted tags:")
    for tag in tags:
        confidence_bar = "█" * int(tag.confidence * 20)
        print(f"    [{tag.category}] {tag.name}: {confidence_bar} ({tag.confidence:.1%})")


def main():
    if len(sys.argv) < 2:
        print("Usage: python basic_analysis.py <audio_or_video_file>")
        print("\nSupported formats: WAV, MP3, FLAC, MP4, MKV, etc.")
        sys.exit(1)

    file_path = sys.argv[1]

    try:
        # Run all analyses
        analyze_audio_file(file_path)
        generate_fingerprint(file_path)
        auto_tag_content(file_path)

    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: crates/psm-player-frequency/examples/basic_analysis.rs
//! Basic Frequency Analysis Example
//!
//! Demonstrates how to use the psm-player-frequency crate for audio analysis.
//!
//! # Usage
//! ```bash
//! cargo run --example basic_analysis -- path/to/audio.wav
//! ```

use anyhow::Result;
use psm_player_frequency::{AudioData, FrequencyAnalyzer};
use std::env;

fn main() -> Result<()> {
    // Initialize logging
    tracing_subscriber::fmt::init();

    // Get file path from arguments
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <audio_file>", args[0]);
        eprintln!("\nSupported formats: WAV");
        std::process::exit(1);
    }

    let file_path = &args[1];
    println!("Analyzing: {}", file_path);
    println!("{}", "-".repeat(50));

    // Load audio file
    let audio = load_wav(file_path)?;
    println!("Loaded {} samples at {}Hz", audio.samples.len(), audio.sample_rate);

    // Create analyzer
    let analyzer = FrequencyAnalyzer::new(4096, 2048);

    // Get dominant frequencies
    println!("\nDominant Frequencies:");
    let dominant = analyzer.dominant_frequencies(&audio.samples, audio.sample_rate, 10)?;
    for (i, freq) in dominant.iter().enumerate() {
        println!(
            "  {}. {:>8.1} Hz  (magnitude: {:.4})",
            i + 1,
            freq.frequency_hz,
            freq.magnitude
        );
    }

    // Compute frequency signature
    println!("\nFrequency Signature:");
    let signature = analyzer.compute_signature(&audio.samples, audio.sample_rate)?;

    println!("  Feature vector: {} dimensions", signature.features.len());
    println!("  Centroid:       {:.1} Hz", signature.centroid);
    println!("  Flatness:       {:.4}", signature.flatness);

    // Full analysis
    println!("\nFull Analysis:");
    let analysis = analyzer.analyze(&audio.samples, audio.sample_rate)?;

    println!("  Spectral centroid:  {:.1} Hz", analysis.spectral_centroid);
    println!("  Spectral rolloff:   {:.1} Hz", analysis.spectral_rolloff);
    println!("  Spectral flatness:  {:.4}", analysis.spectral_flatness);
    println!("  Zero crossing rate: {:.4}", analysis.zero_crossing_rate);

    println!("\nBand Energies:");
    println!("  Sub-bass (20-60 Hz):    {:.4}", analysis.band_energies.sub_bass);
    println!("  Bass (60-250 Hz):       {:.4}", analysis.band_energies.bass);
    println!("  Low-mid (250-500 Hz):   {:.4}", analysis.band_energies.low_mid);
    println!("  Mid (500-2000 Hz):      {:.4}", analysis.band_energies.mid);
    println!("  High-mid (2-4 kHz):     {:.4}", analysis.band_energies.high_mid);
    println!("  High (4-20 kHz):        {:.4}", analysis.band_energies.high);

    Ok(())
}

/// Load a WAV file into AudioData
fn load_wav(path: &str) -> Result<AudioData> {
    let reader = hound::WavReader::open(path)?;
    let spec = reader.spec();

    let samples: Vec<f32> = reader
        .into_samples::<i16>()
        .filter_map(|s| s.ok())
        .map(|s| s as f32 / 32768.0)
        .collect();

    let duration_secs = samples.len() as f64 / spec.sample_rate as f64;

    Ok(AudioData {
        samples,
        sample_rate: spec.sample_rate,
        channels: spec.channels as u32,
        duration_secs,
    })
}
